src/adapter/src/optimize/mod.rs

```diff
@@ -0,0 +1,160 @@
+//!     code cannot first execute some optimization stages for one type and then
+//!     some stages for a different type.
+//!   - The only way to construct such a struct is by running the [`Optimize`]
+//!     stage that produces it. This ensures that client code cannot interfere
+//!     with the pipeline.
+//!   - In general, the internals of these structs can be accessed only behind a
+//!     shared reference. This ensures that client code can look up information
+//!     from intermediate stages but cannot modify it.
+//!   - Timestamp selection is modeled as a conversion between structs that are
+//!     adjacent in the pipeline.
+//!   - The struct representing the result of the final stage of the
+//!     optimization pipeline can be destructed to access its internals.
+//! - The `Send + Sync` trait bounds on the `Self` and `From` types ensure that
+//!   [`Optimize`] instances can be passed to different threads.
+//!
+//! For details, see the `20230714_optimizer_interface.md` design doc in this
+//! repository.
+
+mod materialized_view;
+
+// Re-export optimzier structs
+pub use materialized_view::OptimizeMaterializedView;
+
+use mz_compute_types::dataflows::DataflowDescription;
+use mz_compute_types::plan::Plan;
+use mz_expr::OptimizedMirRelationExpr;
+use mz_sql::plan::PlanError;
+use mz_sql::session::vars::SystemVars;
+use mz_transform::TransformError;
+
+use crate::AdapterError;
+
+/// A trait that represents an optimization stage.
+///
+/// The trait is implemented by structs that encapsulate the context needed to
+/// run an end-to-end optimization pipeline for a specific statement type
+/// (`Index`, `View`, `MaterializedView`, `Subscribe`, `Select`).
+///
+/// Each implementation represents a concrete optimization stage for a fixed
+/// statement type that consumes an input of type `From` and produces output of
+/// type `Self::To`.
+///
+/// The generic lifetime `'ctx` models the lifetime of the optimizer context and
+/// can be passed to the optimizer struct and the `Self::To` types.
+///
+/// The `'s: 'ctx` bound in the `optimize` method call ensures that an optimizer
+/// instance can run an optimization stage that produces a `Self::To` with
+/// `&'ctx` references.
+pub trait Optimize<'ctx, From>: Send + Sync
+where
+    From: Send + Sync,
+{
+    type To: Send + Sync + 'ctx;
+
+    /// Execute the optimization stage, transforming the input plan of type
+    /// `From` to an output plan of type `To`.
+    fn optimize<'s: 'ctx>(&'s mut self, plan: From) -> Result<Self::To, OptimizerError>;
+
+    /// Execute the optimization stage and panic if an error occurs.
+    ///
+    /// See [`Optimize::optimize`].
+    fn must_optimize<'s: 'ctx>(&'s mut self, expr: From) -> Self::To {
+        match self.optimize(expr) {
+            Ok(ok) => ok,
+            Err(err) => panic!("must_optimize call failed: {err}"),
+        }
+    }
+}
+
+// Feature flags for the optimizer.
+pub struct OptimizerFlags {
```

Comment:
Nit: "Feature flags" is a bit misleading here, as this also contains system config options that are not flags (`persist_fast_path_limit`). Maybe rename this to `OptimizerOptions`?

Reply:
I'll rename this to `OptimizerConfig` before merging.